File: test/language/parameterized_types/java/parameterized_types/packed_array_element_param/PackedArrayElementParamTest.java
package parameterized_types.packed_array_element_param;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.io.File;
import java.math.BigInteger;

import zserio.runtime.ZserioError;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;
import zserio.runtime.io.FileBitStreamWriter;

public class PackedArrayElementParamTest
{
    @Test
    public void bitSizeOfLength1() throws IOException, ZserioError
    {
        checkBitSizeOf(NUM_BLOCKS1);
    }

    @Test
    public void bitSizeOfLength2() throws IOException, ZserioError
    {
        checkBitSizeOf(NUM_BLOCK2);
    }

    @Test
    public void bitSizeOfLength3() throws IOException, ZserioError
    {
        checkBitSizeOf(NUM_BLOCK3);
    }

    @Test
    public void writeReadLength1() throws IOException, ZserioError
    {
        checkWriteRead(NUM_BLOCKS1);
    }

    @Test
    public void writeReadLength2() throws IOException, ZserioError
    {
        checkWriteRead(NUM_BLOCK2);
    }

    @Test
    public void writeReadLength3() throws IOException, ZserioError
    {
        checkWriteRead(NUM_BLOCK3);
    }

    @Test
    public void writeReadFileLength1() throws IOException, ZserioError
    {
        checkWriteReadFile(NUM_BLOCKS1);
    }

    @Test
    public void writeReadFileLength2() throws IOException, ZserioError
    {
        checkWriteReadFile(NUM_BLOCK2);
    }

    @Test
    public void writeReadFileLength3() throws IOException, ZserioError
    {
        checkWriteReadFile(NUM_BLOCK3);
    }

    private void checkBitSizeOf(int numBlocks) throws IOException, ZserioError
    {
        final Database database = createDatabase(numBlocks);
        final int unpackedBitsizeOf = calcUnpackedDatabaseBitSize(numBlocks);
        final int packedBitsizeOf = database.bitSizeOf();
        final double minCompressionRatio = 0.12;

        assertTrue(unpackedBitsizeOf * minCompressionRatio > packedBitsizeOf, () ->
                "Unpacked array has " + unpackedBitsizeOf + " bits, packed array has " + packedBitsizeOf +
                " bits, " + "compression ratio is " + packedBitsizeOf * 100.0 / unpackedBitsizeOf + "%!");
    }

    private void checkWriteRead(int numBlocks) throws IOException, ZserioError
    {
        final Database database = createDatabase(numBlocks);
        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        database.write(writer);
        writer.close();

        final long writtenBitPosition = writer.getBitPosition();
        assertEquals(database.bitSizeOf(), writtenBitPosition);
        assertEquals(database.initializeOffsets(), writtenBitPosition);

        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(writer.toByteArray());
        final Database readDatabase = new Database(reader);
        assertEquals(database, readDatabase);
    }

    private void checkWriteReadFile(int numBlocks) throws IOException, ZserioError
    {
        final Database database = createDatabase(numBlocks);
        final File file = new File(BLOB_NAME_BASE + numBlocks + ".blob");
        final FileBitStreamWriter writer = new FileBitStreamWriter(file);
        database.write(writer);
        writer.close();

        final Database readDatabase = new Database(file);
        assertEquals(database, readDatabase);
    }

    private Database createDatabase(int numBlocks)
    {
        final BlockHeader[] blockHeaders = new BlockHeader[numBlocks];
        final Block[] blocks = new Block[numBlocks];
        for (int i = 0; i < numBlocks; ++i)
        {
            final int numItems = i + 1;
            final BlockHeader blockHeader = new BlockHeader(numItems, 0);
            blockHeaders[i] = blockHeader;
            final long[] items = new long[numItems];
            for (int j = 0; j < numItems; ++j)
                items[j] = j * 2;

            blocks[i] = new Block(blockHeader, BigInteger.valueOf(numItems), items);
        }

        return new Database(numBlocks, blockHeaders, blocks);
    }

    private int calcUnpackedDatabaseBitSize(int numBlocks)
    {
        int bitSize = 16; // numBlocks
        bitSize += numBlocks * (16 + 32); // headers
        for (int i = 0; i < numBlocks; ++i)
        {
            final int numItems = i + 1;
            bitSize += 64 + numItems * 64; // blocks[i]
        }

        return bitSize;
    }

    private static final String BLOB_NAME_BASE = "packed_array_element_param_";
    private static final int NUM_BLOCKS1 = 50;
    private static final int NUM_BLOCK2 = 100;
    private static final int NUM_BLOCK3 = 1000;
}

File: test/extensions/language/parameterized_types/python/PackedArrayElementParamTest.py
import os
import zserio

import ParameterizedTypes
from testutils import getApiDir


class PackedArrayElementParamTest(ParameterizedTypes.TestCase):
    def testBitSizeOfLength1(self):
        self._checkBitSizeOf(self.NUM_BLOCKS1)

    def testBitSizeOfLength2(self):
        self._checkBitSizeOf(self.NUM_BLOCKS2)

    def testBitSizeOfLength3(self):
        self._checkBitSizeOf(self.NUM_BLOCKS3)

    def testWriteReadLength1(self):
        self._checkWriteRead(self.NUM_BLOCKS1)

    def testWriteReadLength2(self):
        self._checkWriteRead(self.NUM_BLOCKS2)

    def testWriteReadLength3(self):
        self._checkWriteRead(self.NUM_BLOCKS3)

    def testWriteReadFileLength1(self):
        self._checkWriteReadFile(self.NUM_BLOCKS1)

    def testWriteReadFileLength2(self):
        self._checkWriteReadFile(self.NUM_BLOCKS2)

    def testWriteReadFileLength3(self):
        self._checkWriteReadFile(self.NUM_BLOCKS3)

    def _checkBitSizeOf(self, numBlocks):
        database = self._createDatabase(numBlocks)
        unpackedBitsizeOf = PackedArrayElementParamTest._calcUnpackedDatabaseBitSize(numBlocks)
        packedBitsizeOf = database.bitsizeof()
        minCompressionRatio = 0.12
        self.assertTrue(
            unpackedBitsizeOf * minCompressionRatio > packedBitsizeOf,
            "Unpacked array has "
            + str(unpackedBitsizeOf)
            + " bits, packed array has "
            + str(packedBitsizeOf)
            + " bits, "
            + "compression ratio is "
            + str(packedBitsizeOf / unpackedBitsizeOf * 100)
            + "%!",
        )

    def _checkWriteRead(self, numBlocks):
        database = self._createDatabase(numBlocks)
        writer = zserio.BitStreamWriter()
        database.initialize_offsets(writer.bitposition)
        database.write(writer)
        self.assertEqual(database.bitsizeof(), writer.bitposition)
        self.assertEqual(database.initialize_offsets(), writer.bitposition)

        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readDatabase = self.api.Database.from_reader(reader)
        self.assertEqual(database, readDatabase)

    def _checkWriteReadFile(self, numBlocks):
        database = self._createDatabase(numBlocks)
        filename = self.BLOB_NAME_BASE + str(numBlocks) + ".blob"
        zserio.serialize_to_file(database, filename)

        readDatabase = zserio.deserialize_from_file(self.api.Database, filename)
        self.assertEqual(database, readDatabase)

    def _createDatabase(self, numBlocks):
        blockHeaderList = []
        blockList = []
        for i in range(numBlocks):
            numItems = i + 1
            blockHeader = self.api.BlockHeader(numItems, 0)
            blockHeaderList.append(blockHeader)
            itemList = []
            for j in range(numItems):
                itemList.append(j * 2)

            blockList.append(self.api.Block(blockHeader, numItems, itemList))

        return self.api.Database(numBlocks, blockHeaderList, blockList)

    @staticmethod
    def _calcUnpackedDatabaseBitSize(numBlocks):
        bitSize = 16  # numBlocks
        bitSize += numBlocks * (16 + 32)  # headers
        for i in range(numBlocks):
            numItems = i + 1
            bitSize += 64 + numItems * 64  # blocks[i]

        return bitSize

    BLOB_NAME_BASE = os.path.join(getApiDir(os.path.dirname(__file__)), "packed_array_element_param_")
    NUM_BLOCKS1 = 50
    NUM_BLOCKS2 = 100
    NUM_BLOCKS3 = 1000
